source/simulator/core.cc

```diff
@@ -755,20 +776,75 @@ namespace aspect
+      reset_outflow_boundary_ids(boundary_id_offset);
+
     // let plugins add more constraints if they so choose, then close the
     // constraints object
-    signals.post_constraints_creation(*this, current_constraints);
+    signals.post_constraints_creation(*this, new_current_constraints);
 
-    current_constraints.close();
+    new_current_constraints.close();
 
     // let the melt handler add its constraints once before we solve the porosity system for the first time
     if (parameters.include_melt_transport)
-      melt_handler->save_constraints (current_constraints);
+      melt_handler->save_constraints (new_current_constraints);
+
+    // Now check if the current_constraints we just computed changed from before. Do this before the melt handler
+    // adds constraints for the melt cells, because they are allowed to change (and often do) without us having
+    // to reinit the sparsity pattern.
+    bool constraints_changed = false;
+
+    // If the mesh got refined and the size of the linear system changed, the old and new constraint
+    // matrices will have different entries, and we can not easily compare them.
+    // However, in case of mesh refinement we have to rebuild the matrix anyway, so we can skip the
+    // step that checks if constraints have changed.
+    bool mesh_has_changed = false;
+    if (!(current_constraints.get_local_lines().size() == new_current_constraints.get_local_lines().size()))
+      mesh_has_changed = true;
+    else if (!(current_constraints.get_local_lines() == new_current_constraints.get_local_lines()))
+      mesh_has_changed = true;
+
+    if (!mesh_has_changed)
+      {
+#if DEAL_II_VERSION_GTE(9,0,0)
+        for (auto &row: current_constraints.get_lines())
+          {
+            if (!new_current_constraints.is_constrained(row.index))
+              {
+                constraints_changed = true;
+                break;
+              }
+          }
+#else
+        for (const auto row: current_constraints.get_local_lines())
+          {
+            if (current_constraints.is_constrained(row)
+                != new_current_constraints.is_constrained(row))
+              {
+                constraints_changed = true;
+                break;
+              }
+          }
+#endif
+
+        const bool any_constraints_changed = Utilities::MPI::sum(constraints_changed ? 1 : 0,
+                                                                 mpi_communicator) > 0;
+        if (any_constraints_changed)
+          rebuild_sparsity_and_matrices = true;
+      }
+
+#if DEAL_II_VERSION_GTE(9,0,0)
+    current_constraints.copy_from(new_current_constraints);
+#else
+    current_constraints.clear ();
+    current_constraints.reinit (introspection.index_sets.system_relevant_set);
+    current_constraints.merge (new_current_constraints);
+#endif
 
     if (time_step == 0 && parameters.include_melt_transport)
       melt_handler->add_current_constraints (current_constraints);
```

Comment:
I am worried that the melt handler will break our logic: if we add constraints here, we will rebuild the sparsity pattern even though that shouldn't be necessary. I am not quite sure how to fix this. Maybe ignore p_c indices in the check above?!